warplib/src/test/warp_memory_test.cairo

```diff
@@ -0,0 +1,307 @@
+
+    warp_memory.alloc(10);
+    assert(warp_memory.pointer == 15, 'Pointer should be 15');
+}
+
+#[test]
+#[should_panic]
+fn test_alloc_overflow(){
+    let mut warp_memory = WarpMemoryTrait::initialize();
+
+    warp_memory.alloc(10);
+
+    // this should crash due to pointer overflow
+    warp_memory.alloc(-1);
+
+}
+
+#[test]
+fn test_unsafe_access() {
+    let mut warp_memory = WarpMemoryTrait::initialize();
+    warp_memory.unsafe_write(3, 5); 
+    
+    let read_from_unallocated = warp_memory.unsafe_read(3);
+    assert(read_from_unallocated == 5, 'Invalid read value');
+    assert(warp_memory.pointer == 0, 'Pointer should be 0');
+}
+
+#[test] 
+#[should_panic]
+fn test_unallocated_write() {
+    let mut warp_memory = WarpMemoryTrait::initialize();
+    warp_memory.write(3, 5); 
+}
+
+#[test] 
+#[should_panic]
+fn test_unallocated_read() {
+    let mut warp_memory = WarpMemoryTrait::initialize();
+    warp_memory.read(3); 
+}
+
+#[test]
+fn test_access(){
+    let mut warp_memory = WarpMemoryTrait::initialize();
+
+    let ptr1 = warp_memory.unsafe_alloc(2);
+    assert(ptr1 == 0, 'Pointer is not 0');
+    assert(warp_memory.pointer == 2, 'Pointer is not 2');
+
+    warp_memory.write(ptr1, 2); 
+    warp_memory.write(ptr1 + 1, 3); 
+    assert(warp_memory.read(ptr1) == 2, 'Invalid read value');
+    assert(warp_memory.read(ptr1 + 1) == 3, 'Invalid read value');
+
+    let ptr2 = warp_memory.unsafe_alloc(3);
+    assert(ptr2 == 2, 'Pointer is not 2');
+    assert(warp_memory.pointer == 5, 'Pointer is not 5');
+
+    warp_memory.write(ptr2, 4); 
+    warp_memory.write(ptr2 + 1, 5); 
+    warp_memory.write(ptr2 + 2, 6); 
+    assert(warp_memory.read(ptr2) == 4, 'Invalid read value');
+    assert(warp_memory.read(ptr2 + 1) == 5, 'Invalid read value');
+    assert(warp_memory.read(ptr2 + 2) == 6, 'Invalid read value');
+}
+
+// ==================== Complex Functions ====================
+use warplib::warp_memory::WarpMemoryArraysTrait;
+
+#[test]
+fn test_get_or_create_id(){
+    let mut warp_memory = WarpMemoryTrait::initialize();
+
+    let ptr1 = warp_memory.unsafe_alloc(2);
+    warp_memory.write(ptr1, 3);
+
+    let read_id1 = warp_memory.get_or_create_id(ptr1, 10);
+    assert(read_id1 == 3, 'Invalid read id value');
+
+    let read_id2 = warp_memory.get_or_create_id(ptr1 + 1, 10);
+    assert(read_id2 == 2, 'Invalid read id value');
+    assert(warp_memory.pointer == 12, 'Invalid pointer value');
+}
+
+#[test]
+fn test_new_dynamic_array(){
+    let mut warp_memory = WarpMemoryTrait::initialize();
+
+    let dyn_array = warp_memory.new_dynamic_array(-2, 1);
+    assert(warp_memory.pointer == -1, 'Invalid pointer value');
+    assert(warp_memory.read(dyn_array) == -2, 'Invalid length value');
+}
+
+#[test]
+#[should_panic]
+fn test_new_dynamic_array_overflow_1(){
+    let mut warp_memory = WarpMemoryTrait::initialize();
+
+    let dyn_array = warp_memory.new_dynamic_array(-1, 1);
```

Comment:
If you assign `-1` to a constant named `MAX_FELT252` it will be much more readable